File: fortran/src/imc_source.f90
subroutine imc_get_energy_sources(e_surf, e_body, e_total)

    use imc_global_mesh_data, only: ncells, fleck, sigma_p, temp, dx
    use imc_global_bcon_data, only: t0
    use imc_global_phys_data, only: sb, a, c
    use imc_global_time_data, only: dt

    implicit none

    real(kind=8), intent(out)                    :: e_surf
    real(kind=8), intent(out)                    :: e_total
    real(kind=8), dimension(ncells), intent(out) :: e_body

    ! Left-hand boundary is a black-body emitter at constant temperature T0
    ! (Energy radiated per unit area = sigma.T0**4 (sigma = S-B constant)
    e_surf = sb * t0**4 * dt

    ! Emission source term
    e_body(:) = (fleck(:) * sigma_p(:) * a * c * temp(:)**4 * dx * dt)

    ! Total energy emitted
    e_total = e_surf + sum(e_body(:))

    write(*,'(/,A)') 'Energy radiated in timestep:'
    write(*,'(/,A)') 'In total:'
    write(*,'(ES24.16)') e_total
    write(*,'(/,A)') 'By boundary condition:'
    write(*,'(ES24.16)') e_surf
    write(*,'(/,A)') 'By mesh:'
    write(*,'(4(ES16.8))') e_body

end subroutine imc_get_energy_sources


subroutine imc_get_emission_probabilities(e_surf, e_body, e_total, p_surf, p_body)

    use imc_global_mesh_data, only: ncells

    implicit none

    real(kind=8), intent(in)                       :: e_surf
    real(kind=8), intent(in)                       :: e_total
    real(kind=8), dimension(ncells), intent(in)    :: e_body
    real(kind=8), intent(out)                      :: p_surf
    real(kind=8), dimension(0:ncells), intent(out) :: p_body
    integer                                        :: jcell

    p_surf = e_surf / e_total

    ! Probability of each cell given that the particle is from the mesh not the surface

    p_body(0) = 0.d0
    do jcell = 1, ncells
      p_body(jcell) = p_body(jcell-1) + e_body(jcell) / sum(e_body(:))
    enddo

    write(*,'(A)')
    write(*,'(A)') 'Probabilities:'
    write(*,'(A)')
    write(*,'(A)') 'In total'
    write(*,'(ES16.8)') p_surf + sum(p_body(:))
    write(*,'(A)')
    write(*,'(A)') 'Of boundary particle'
    write(*,'(ES16.8)') p_surf
    write(*,'(A)')
    write(*,'(A)') 'Of mesh particle (cumulative)'
    write(*,'(10F4.1)') p_body(1:ncells)

end subroutine imc_get_emission_probabilities


subroutine imc_get_source_particle_numbers(p_surf, p_body, n_surf, n_body)

    !use imc_global_bcon_data, only: t0
    use imc_global_mesh_data, only: ncells !, dx, temp, nodepos
    use imc_global_part_data, only: n_max, n_census, n_input !, , dead, xpos, icell, mu, &
    !                              & frq, nrg, startnrg, t, origin, inext
    !use imc_global_phys_data, only: h
    !use imc_global_time_data, only: step, dt, time
    !use imc_utils,            only: samplePlanckspectrum

    implicit none

    real(kind=8), intent(in)                        :: p_surf
    real(kind=8), dimension(0:ncells), intent(in)   :: p_body
    integer(kind=8), intent(out)                    :: n_surf
    integer(kind=8), dimension(ncells), intent(out) :: n_body
    real(kind=8)                            :: eta
    integer(kind=8)                         :: ip
    integer(kind=8)                         :: irannum
    integer(kind=8)                         :: n_source
    real(kind=8), allocatable, dimension(:) :: r
    integer                                 :: jcell

    ! Determine total number of particles to source this time-step
    n_source = n_input
    if ((n_source + n_census) > n_max) &
        n_source = n_max - n_census - ncells - 1

    write(*,'(/,A,I8,A)') 'Sourcing ',n_source,' particles this timestep'
    write(*,'(A,I8,A)') '(User requested ',n_input,' per timestep)'

    n_surf    = 0
    n_body(:) = 0

    allocate(r(2*n_source))
    call random_number(r)
    irannum = 0

    ! Calculate the number of particles emitted by the surface source and each mesh cell
    do ip = 1, n_source
        irannum = irannum + 1
        if (r(irannum) <= p_surf) then
            n_surf = n_surf + 1
        else
            irannum = irannum + 1
            eta = r(irannum)
            do jcell = 1, ncells
                if (eta <= p_body(jcell)) then
                    n_body(jcell) = n_body(jcell) + 1
                    exit
                endif
            enddo
        endif
    enddo
    deallocate(r)

    write(*,'(/,A)') 'Body source'
    write(*,'(10I6)') n_body(:)

end subroutine imc_get_source_particle_numbers


subroutine imc_source_particles(e_surf, e_body, n_surf, n_body)

    use imc_global_bcon_data, only: t0
    use imc_global_mesh_data, only: ncells, dx, temp, nodepos
    use imc_global_part_data, only: icell, inext, n_max, nrg, startnrg, dead, xpos, icell, mu, &
                                  & frq, nrg, startnrg, t, origin, inext
    use imc_global_phys_data, only: h
    use imc_global_time_data, only: dt, time ! step, dt, time
    use imc_utils,            only: samplePlanckspectrum

    implicit none

    real(kind=8), intent(in)                       :: e_surf
    real(kind=8), dimension(ncells), intent(in)    :: e_body
    integer(kind=8), intent(in)                    :: n_surf
    integer(kind=8), dimension(ncells), intent(in) :: n_body
    integer(kind=8)                         :: ip
    integer(kind=8)                         :: irannum
    real(kind=8), allocatable, dimension(:) :: r
    integer                                 :: jcell
    real(kind=8)                            :: jnrg
    real(kind=8)                            :: jstartnrg

    allocate(r(2*n_surf + 4*sum(n_body(:))))
    call random_number(r)
    irannum = 0

    ! Create the surface-source particles
    jnrg = e_surf/real(n_surf, 8)
    jstartnrg = jnrg
    do ip = 1, n_surf
        do while (.not. dead(inext))
           inext = inext + 1
           if (inext > n_max) inext = 1
        enddo
        dead(inext)     = .false.
        origin(inext)   = -1
        xpos(inext)     = 0.d0
        irannum         = irannum + 1
        t(inext)        = time + r(irannum) * dt
        irannum         = irannum + 1
        mu(inext)       = sqrt(r(irannum))                    ! Corresponds to f(mu) = 2mu
        frq(inext)      = t0 * samplePlanckspectrum() / h
        icell(inext)    = 1
        nrg(inext)      = jnrg
        startnrg(inext) = jstartnrg
    enddo

    ! Create the body-source particles
    do jcell = 1, ncells
        if (n_body(jcell) <= 0) &
            cycle
        jnrg = e_body(jcell)/real(n_body(jcell), 8)
        jstartnrg = jnrg
        do ip = 1, n_body(jcell)
            do while (.not. dead(inext))
                inext = inext + 1
                if (inext > n_max) inext = 1
            enddo
            dead(inext)     = .false.
            origin(inext)   = jcell
            irannum         = irannum + 1
            xpos(inext)     = nodepos(jcell) + r(irannum) * dx
            irannum         = irannum + 1
            mu(inext)       = 1.d0 - 2.d0 * r(irannum)
            irannum         = irannum + 1
            t(inext)        = time + r(irannum) * dt
            irannum         = irannum + 1
            frq(inext)      = temp(jcell) * abs(log(r(irannum))) / h
            icell(inext)    = jcell
            nrg(inext)      = jnrg
            startnrg(inext) = jstartnrg
        enddo
    enddo

    deallocate(r)

end subroutine imc_source_particles


subroutine imc_source

    use imc_global_mesh_data, only: ncells
    use imc_global_part_data, only: n_max, dead
    use imc_global_phys_data, only: h
    use imc_global_time_data, only: step

    implicit none

    real(kind=8)                            :: e_surf
    real(kind=8)                            :: e_total
    real(kind=8), dimension(ncells)         :: e_body
    real(kind=8)                            :: p_surf
    real(kind=8), dimension(0:ncells)       :: p_body
    integer(kind=8)                         :: n_surf
    integer(kind=8), dimension(ncells)      :: n_body
    integer(kind=8)                         :: numparts
    integer(kind=8)                         :: ip

    write(*,'(/,A)')    '-------------------------------------------------------------------------------'
    write(*,'(A,I4,A)') 'Source step (',step,')'
    write(*,'(A)')      '-------------------------------------------------------------------------------'

    ! Determine probability of particles belonging to sources
    ! -------------------------------------------------------

    ! Get the energy source terms
    call imc_get_energy_sources(e_surf, e_body, e_total)

    ! Emission probabilities
    call imc_get_emission_probabilities(e_surf, e_body, e_total, p_surf, p_body)

    ! Number of source particles
    call imc_get_source_particle_numbers(p_surf, p_body, n_surf, n_body)

    ! Create the particles
    call imc_source_particles(e_surf, e_body, n_surf, n_body)

    ! Particle count
    numparts = 0
    do ip = 1, n_max
        if (.not. dead(ip)) numparts = numparts + 1
    enddo
    write(*,'(A,I12)') 'Number of particles in the system = ', numparts

end subroutine imc_source

File: python/src/imc_source.py
"""Source IMC particles."""

import math
import random as ran

import numpy as np

import imc_global_part_data as part
import imc_global_mesh_data as mesh
import imc_global_phys_data as phys
import imc_global_bcon_data as bcon
import imc_global_time_data as time

from imc_utils import sample_planck_spectrum


def imc_get_energy_sources():
    """Get energy source terms for surface and mesh."""
    # Left-hand boundary is a black-body emitter at constant temperature T0
    # (Energy radiatied per unit area = sigma.T0**4 (sigma = S-B constant)
    e_surf = phys.sb * bcon.T0 ** 4 * time.dt

    # Emission source term
    e_body = np.zeros(mesh.ncells)  # Energy emitted per cell per time-step
    e_body[:] = (
        mesh.fleck[:]
        * mesh.sigma_p[:]
        * phys.a
        * phys.c
        * mesh.temp[:] ** 4
        * mesh.dx
        * time.dt
    )

    # Total energy emitted
    e_total = e_surf + sum(e_body[:])

    print("\nEnergy radiated in timestep:")
    print("\nIn total:")
    print("{:24.16E}".format(e_total))
    print("\nBy boundary condition:")
    print("{:24.16E}".format(e_surf))
    print("\nBy mesh:")
    print(e_body)

    return e_surf, e_body, e_total


def imc_get_emission_probabilities(e_surf, e_body, e_total):
    """Convert energy source terms to particle emission probabilities."""
    p_surf = e_surf / e_total
    # Probability of each cell _given that the particle is from the mesh not the
    # surface_
    p_body = np.zeros(mesh.ncells)
    p_body[:] = np.cumsum(e_body[:]) / sum(e_body[:])

    print("\nProbabilities:")
    print("\nIn total")
    print("{:16.8E}".format(p_surf + sum(p_body[:])))
    print("\nOf boundary particle")
    print("{:16.8E}".format(p_surf))
    print("\nOf mesh particle (cumulative)")
    print(p_body[:])

    return p_surf, p_body


def imc_get_source_particle_numbers(p_surf, p_body):
    """Calculate number of source particles to create at surface / throughout mesh."""
    n_census = part.n_census
    n_input = part.n_input
    n_max = part.n_max

    # Determine total number of particles to source this time-step
    n_source = n_input
    if (n_source + n_census) > n_max:
        n_source = n_max - n_census - mesh.ncells - 1

    print("\nSourcing {:8d} particles this timestep".format(n_source))
    print("(User requested {:8d} per timestep)".format(n_input))

    n_surf = 0
    n_body = np.zeros(mesh.ncells, dtype=np.uint64)

    # Calculate the number of particles emitted by the surface souce and each mesh cell
    for _ in range(n_source):
        if ran.random() <= p_surf:
            n_surf += 1
        else:
            eta = ran.random()
            for icell in range(mesh.ncells):
                if eta <= p_body[icell]:
                    n_body[icell] += 1
                    break

    print("\nBody source")
    print(n_body)

    return n_surf, n_body


def imc_source_particles(e_surf, n_surf, e_body, n_body):
    """For known energy distribution (surface and mesh), create source particles."""
    # Create the surface-source particles
    nrg = e_surf / float(n_surf)
    startnrg = nrg
    for _ in range(n_surf):
        origin = -1
        xpos = 0.0
        muu = math.sqrt(ran.random())  # Corresponds to f(mu) = 2mu
        ttt = time.time + ran.random() * time.dt
        frq = bcon.T0 * sample_planck_spectrum() / phys.h
        part.prop.append(
            [origin, ttt, 0, xpos, muu, frq, nrg, startnrg]
        )  # Add this ptcl to the global list

    # Create the body-source particles
    for icell in range(mesh.ncells):
        if n_body[icell] <= 0:
            continue
        nrg = e_body[icell] / float(n_body[icell])
        startnrg = nrg
        for _ in range(n_body[icell]):
            origin = icell
            xpos = mesh.nodepos[icell] + ran.random() * mesh.dx
            muu = 1.0 - 2.0 * ran.random()
            ttt = time.time + ran.random() * time.dt
            frq = mesh.temp[icell] * abs(math.log(ran.random())) / phys.h
            # Add this ptcl to the global list
            part.prop.append([origin, ttt, icell, xpos, muu, frq, nrg, startnrg])


def run():
    """
    Source new IMC particles.

    This routine calculates the energy sources for
    the (a) left-hand boundary (which is currently held at a constant
    temperature, T0), and (b) the computational cells, as well as the overall
    total for the time-step. These are then converted into particle emission
    probabilities. The number of particles to source in this time-step is
    determined (ensuring that the total number in the system does not exceed
    some pre-defined maximum), and then these are attributed either to the
    boundary or to one of the mesh cells, according to the probabilities
    calculated earlier. The particles are then created.
    """
    print("\n" + "-" * 79)
    print("Source step ({:4d})".format(time.step))
    print("-" * 79)

    # Determine probability of particles belonging to sources
    # -------------------------------------------------------

    # Get the energy source terms
    (e_surf, e_body, e_total) = imc_get_energy_sources()

    # Emission probabilities
    (p_surf, p_body) = imc_get_emission_probabilities(e_surf, e_body, e_total)

    # Number of source particles
    (n_surf, n_body) = imc_get_source_particle_numbers(p_surf, p_body)

    # Create the particles
    imc_source_particles(e_surf, n_surf, e_body, n_body)

    # Particle count
    print("Number of particles in the system = {:12d}".format(len(part.prop)))
